File: cpp/day-07.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <ranges>
#include <unordered_map>
#include <sstream>
#include <utility>
#include <regex>
#include <stdint.h>
#include <tuple>
#include <stack>
#include <queue>
#include <functional>

auto parse(const std::vector<std::string>& input) {
    std::vector<std::pair<int64_t, std::vector<int64_t>>> equations;
    for (auto& eq : input) {
        auto start = eq.find_first_not_of(' ');
        auto end = eq.find(':');
        auto result = std::stoll(eq.substr(start, end - start));
        end += 1;
        std::vector<int64_t> operands;
        while ((start = eq.find_first_not_of(' ', end)) != eq.npos) {
            end = eq.find(' ', start);
            operands.push_back(std::stoll(eq.substr(start, end - start)));
        }
        equations.push_back(make_pair(result, operands));
    }
    return equations;
}

auto part_one(const std::vector<std::string>& input) {
    int64_t ans = 0;
    auto equations = parse(input);
    for (auto& [result, operands] : equations) {
        if (operands.empty()) continue;
        std::function<bool (int64_t, size_t)> possible = [&](int64_t curr, size_t index) -> bool {
            if (index == operands.size()) return curr == result;
            bool can = false;
            can |= possible(curr + operands[index], index + 1);
            can |= possible(curr * operands[index], index + 1);
            return can;
        };
        if (possible(operands[0], 1)) ans += result;
    }
    return ans;
}

auto part_two(const std::vector<std::string>& input) {
    int64_t ans = 0;
    auto equations = parse(input);
    for (auto& [result, operands] : equations) {
        if (operands.empty()) continue;
        std::function<bool (int64_t, size_t)> possible = [&](int64_t curr, size_t index) -> bool {
            if (index == operands.size()) return curr == result;
            bool can = false;
            can |= possible(curr + operands[index], index + 1);
            can |= possible(curr * operands[index], index + 1);
            can |= possible(std::stoll(std::to_string(curr) + std::to_string(operands[index])), index + 1);
            return can;
        };
        if (possible(operands[0], 1)) ans += result;
    }
    return ans;
}

auto main() -> int {
    std::vector<std::string> input;
    std::string line;
    while (std::getline(std::cin, line)) {
        input.push_back(std::move(line));
    }

    std::cout << "Part 1: " << part_one(input) << std::endl;
    std::cout << "Part 2: " << part_two(input) << std::endl;

    return 0;
}

File: rust/src/bin/day-07.rs

fn main() {
    let input = std::io::stdin()
        .lines()
        .filter_map(Result::ok)
        .collect::<Vec<_>>();

    println!("Part 1: {}", part_one(&input));
    println!("Part 1: {}", part_two(&input));
}

fn parse(input: &Vec<String>) -> Vec<(i64, Vec<i64>)> {
    let mut equations: Vec<(i64, Vec<i64>)> = Vec::new();
    for eq in input {
        let mut it= eq.split(':');
        let result = it.next().unwrap()
            .parse::<i64>().unwrap();
        let operands = it.next().unwrap()
            .split_whitespace()
            .map(|x| x.parse::<i64>().unwrap())
            .collect::<Vec<_>>();
        equations.push((result, operands));
    }
    return equations;
}

fn part_one(input: &Vec<String>) -> i64 {
    let mut ans = 0;
    let equations = parse(input);

    fn possible(result: &i64, operands: &Vec<i64>, curr: i64, index: usize) -> bool {
        if index == operands.len() {
            return *result == curr;
        }
        let mut can = false;
        can |=  possible(result, operands, curr + operands[index], index + 1);
        can |=  possible(result, operands, curr * operands[index], index + 1);
        return can;
    }

    for (result, operands) in equations {
        if operands.is_empty() {
            continue;
        }
        if possible(&result, &operands, operands[0], 1) {
            ans += result;
        }

    }
    return ans;
}

fn part_two(input: &Vec<String>) -> i64 {
    let mut ans = 0;
    let equations = parse(input);

    fn possible(result: &i64, operands: &Vec<i64>, curr: i64, index: usize) -> bool {
        if index == operands.len() {
            return *result == curr;
        }
        let mut can = false;
        can |=  possible(result, operands, curr + operands[index], index + 1);
        can |=  possible(result, operands, curr * operands[index], index + 1);
        can |=  possible(result, operands, (curr.to_string() + &operands[index].to_string()).parse::<i64>().unwrap(), index + 1);
        return can;
    }

    for (result, operands) in equations {
        if operands.is_empty() {
            continue;
        }
        if possible(&result, &operands, operands[0], 1) {
            ans += result;
        }

    }
    return ans;
}